File: Data Structures and Algorithms/DS-A-in-Java/algo/search/jumpSearch/jumpSearch.java
class jumpSearch{
	
	/*
Let’s consider the following array: (0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610). Length of the array is 16. 
Jump search will find the value of 55 with the following steps assuming that the block size to be jumped is 4.
STEP 1: Jump from index 0 to index 4;
STEP 2: Jump from index 4 to index 8;
STEP 3: Jump from index 8 to index 12;
STEP 4: Since the element at index 12 is greater than 55 we will jump back a step to come to index 8.
STEP 5: Perform linear search from index 8 to get the element 55.
	*/
// Assumptions:
// The inputted array can be sorted or unsorted
// Check if the array is sorted or not
// if sorted proceed if not sort it first

	boolean isSorted(int arr[]){
		if(arr.length == 1 || arr.length == 0){
			return true;
		}
		for(int i=1; i<arr.length; i++){
			if(arr[i-1] > arr[i])
				return false;
		}
		return true;
	}

	int toJumpSearch(int[] arr, int size, int key){
		int jump = (int)Math.floor(Math.sqrt(size)); //To get the jump we get the sqrt of the size
		int previous = 0;
        while (arr[Math.min(jump, size)-1] < key) 
        { 
            previous = jump; //Setting the previous jumped index
            jump += (int)Math.floor(Math.sqrt(size)); //Incrementing the jump by the sqrt of size
            if (previous >= size) //If the previous is already > or = to the array size
                return -1; 		 //that means the element is not found so we exit
        } 
        while (arr[previous] < key) 
        { 
            previous++; //This is basically a linear search
  
            
            if (previous == Math.min(jump, size)) //If the previous is = to the minimum of either 
                return -1; 						 //jump or size that means element is ot found
        } 

		if(arr[previous] == key)
			return previous;

		return -1;
	}

	void bubbleSort(int arr[]){
		for(int i=0; i<arr.length-1; i++){
			for(int j=0; j<arr.length-i-1; j++)
				if(arr[j+1] < arr[j]){
					int temp = arr[j+1];
					arr[j+1] = arr[j];
					arr[j] = temp;
				}
		}
	}

	void result(int res){
		if(res==-1)
			System.out.print("Key searched is not found in the array! \n");
		else
			System.out.print("The key is found at index ["+res+"] \n");
	}

	public static void main(String[] args) {
		int arr[] = new int[]{2,1,4,3,6,5,8,7};
		if(!new jumpSearch().isSorted(arr)){
			System.out.print("The array is unsorted, sorting... \n");
			new jumpSearch().bubbleSort(arr);
			int result = new jumpSearch().toJumpSearch(arr, arr.length, 5);
			new jumpSearch().result(result);
		}
		else{
		System.out.print("The array is sorted already \n");
		int result = new jumpSearch().toJumpSearch(arr, arr.length, 5);
		new jumpSearch().result(result);
	}
	}
}

File: Data Structures and Algorithms/DSA-in-C-/algo/searching/jumpSearch/jumpSearch.cpp
#include <iostream>
#include <cmath>

using std::swap;

/*
TODO:
isSorted();   [/]
toSort();     []
jumpSearch(); []
*/

int minimum(int num, int num2){
  int min;
  if(num < num2)
    min = num;
  else if(num2 < num)
    min = num2;
return min;
}

void print(int arr[], int size){
  for(int i=0; i<size; i++)
    std::cout<<"["<<arr[i]<<"] ";
}

bool isSorted(int arr[], int size){
  //Let's check if the array only has 1 element on it. if it is then its true
  //Same goes if the size is 0
  if(size == 1 || size == 0)
    return true;
  for(int i=1; i<size; i++)
    if(arr[i-1] < arr[i]) //Looks for unsorted pair in the array
      return false;   //If there is then it returns false

  return true;
}

void toSort(int arr[], int size){
  for(int i = 0; i<size-1; i++) //First loop that iterates till n-1
    for(int j = 0; j<size-i-1; j++) //Second loop that iterates till n-i-1
        if(arr[j+1] < arr[j])
          swap(arr[j+1] , arr[j]);
}

int jumpSearch(int arr[], int size, int key){
  int jump = (int)sqrt(size);
  int previous = 0;
  while(arr[minimum(jump, size) - 1] < key){
    previous = jump;
    jump += (int)sqrt(size);
    if (previous >= size) //If the previous is already > or = to the array size
                return -1; 		 //that means the element is not found so we exit
   while(arr[previous] < key){
     previous++; //This is basically a linear search
     if (previous == minimum(jump, size)) //If the previous is = to the minimum of either
                return -1; 						 //jump or size that means element is ot found
   }
  }
  if(arr[previous]==key)
    return previous;

  return -1;
}
void result(int res){
  if(res == -1)
    std::cout<<"Key is not found in the array! \n";
  else
    std::cout<<"The key is found at index ["<<res<<"] \n";
}
int main(int argc, char const *argv[]) {
  int size, toSearch;
  std::cout<<"Enter array size: ";std::cin>>size;
  int arr[size];
  for(int i=0; i<size; i++)
    std::cin>>arr[i];
  if(!isSorted(arr, size)){
    toSort(arr, size);
  print(arr, size);
  std::cout<<"[Array wasn't sorted!] \nSearch input: ";std::cin>>toSearch;
  int res = jumpSearch(arr, size, toSearch);
  result(res);
}
else{
  print(arr, size);
  std::cout<<"[Array was sorted!]\nSearch input: ";std::cin>>toSearch;
  int res = jumpSearch(arr, size, toSearch);
  result(res);
}  return 0;
}
